common/canonical_serialization/src/canonical_deserialize.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::prelude::*;
+use std::collections::BTreeMap;
+
+pub trait CanonicalDeserialize {
+    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_canonical_deserializer {
```

Comment:
@sunshowers / @bmwill can you review the macros?  I'm not an expert at Rust macros and want to make sure we have these completely correct